### PR TITLE
rust data loader example

### PR DESCRIPTION
**Pull request by mythmon**
For our recent data loaders, I had prepared two examples, but only had time to demo one of them in the presentation. This is the other one, a data loader example that uses Rust to simulate 10 million poker hands and reports the probability of getting each hand.

### CODE REVIEW

File: examples/loader-rust-to-json/src/data/poker.json.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env rust-script
```
Comment (Fil) [inline examples/loader-rust-to-json/src/data/poker.json.rs:1]: I think this line is not strictly necessary (it's ignored); one could consider it informative ("this script is supposed to run with rust-script"), but in a sense it's also confusing as it means "this script is an executable that will call this particular interpreter", when in fact it is called from framework with a (possibly) different interpreter.

Similarly, it is not necessary to have the `+x` bit set on this file. I would argue that it is (slightly) more secure in general to turn off the executable bit when not strictly required.

In other words, only the `.exe` loaders should have a shebang and be `+x` (and even then, possibly only `+x` for the user or group that runs the script, if possible).

File: examples/loader-rust-to-json/src/data/poker.json.rs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+//! Since Framework uses rust-script, we can define dependencies here.
```
Comment (Fil) [inline examples/loader-rust-to-json/src/data/poker.json.rs:2]: This is great!

File: examples/loader-rust-to-json/src/index.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+})
+```
+
+```js echo
+function formatPercent(x) {
+  // Displays both large and small percentages by adding decimal places
+  // until there is something besides zeroes, or until the limit is hit.
+  for (let decimals = 0; decimals <= 5; decimals++) {
+    const f = d3.format(`.${decimals}%`);
+    let s = f(x);
```
Comment (Fil) [inline examples/loader-rust-to-json/src/index.md:225]:
```suggestion
    const s = f(x);
```

File: examples/loader-rust-to-json/src/index.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+      dx: 2,
+    }),
+  ]
+})
+```
+
+```js echo
+function formatPercent(x) {
+  // Displays both large and small percentages by adding decimal places
+  // until there is something besides zeroes, or until the limit is hit.
```
Comment (Fil) [inline examples/loader-rust-to-json/src/index.md:222]: Neat! Another “trick” I like with percentages is to not round 0.99999 to 100% but display it as `99.9%` (not applicable here, but I can see this function being reused in other contexts).

File: examples/loader-rust-to-json/src/index.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+<div style="max-width: 300px;">
+
+```js
+Inputs.table(hands.summary)
+```
+
+</div>
+
+Taking advantage of Rust’s performance and simple parallelism, the data loader above was able to summarize ${hands.meta.count / 1e6} million hands in ${hands.meta.duration_s.toFixed(1)} seconds.
```
Comment (Fil) [inline examples/loader-rust-to-json/src/index.md:191]: This shows as “the data loader above was able to summarize 10 million hands in 1.3 seconds” and might show as “1 seconds” for someone with a slightly faster machine. You could use milliseconds to avoid having to do a singular “1 second”.

**Comment by Fil** [comment]
Just a few nitpicking comments :) It's a great example!

File: examples/loader-rust-to-json/src/index.md [hunk 6]
```diff
@@ -188,7 +188,7 @@ Inputs.table(hands.summary)
 
 </div>
 
-Taking advantage of Rust’s performance and simple parallelism, the data loader above was able to summarize ${hands.meta.count / 1e6} million hands in ${hands.meta.duration_s.toFixed(1)} seconds.
+Taking advantage of Rust’s performance and simple parallelism, the data loader above was able to summarize ${hands.meta.count / 1e6} million hands in ${hands.meta.duration_ms.toLocaleString()} ms.
```
Comment (mbostock) [inline examples/loader-rust-to-json/src/index.md:191]:
```suggestion
Taking advantage of Rust’s performance and simple parallelism, the data loader above was able to summarize ${hands.meta.count / 1e6} million hands in ${hands.meta.duration_ms.toLocaleString("en-US")} ms.
```

When interpolating localized values into English text, we should specify the locale explicitly. It’ll look weird _e.g._ to interpolate French localization into English.

File: examples/loader-rust-to-json/src/index.md (hunk 5 above)
Comment (mythmon) [reply examples/loader-rust-to-json/src/index.md]: Good call!

**Comment by mythmon** [comment]
As well as the review feedback, I made a fairly mechanical conversion of the Rust data loader into TypeScript, to get a speed comparison. Unsurprisingly, the Rust example is much faster, about 65x faster on my system. 20x of that can be theoretically attributed to parallelism, leaving a ~3x language speed up. I'm not sure if this should be included in the example though. What do you all think?

**Comment by Fil** [comment]
I like the idea!

File: examples/loader-rust-to-json/src/index.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+```js
+const extrapolatedTsTime = handsTs.meta.duration_ms * (hands.meta.count / handsTs.meta.count);
+```
+
+</div>
+
+Taking advantage of Rust’s performance and simple parallelism, the data loader above was able to summarize ${hands.meta.count / 1e6} million hands in ${hands.meta.duration_ms.toLocaleString("en-US")} ms. For comparison, an equivalent data loader written in TypeScript (which you can see in `data/poker-ts.json.ts`) would take ${Math.floor(extrapolatedTsTime / 1000).toLocaleString("en-US")} seconds to do the same simulation.
+
+We can make a quick chart of the poker hands with Plot
```
Comment (allisonhorst) [inline examples/loader-rust-to-json/src/index.md:201]: Tiny nit, add a colon at the end of "We can make a quick chart of the poker hands with Plot"?

File: examples/loader-rust-to-json/src/data/poker.json.rs (hunk 2 above)
Comment (allisonhorst) [reply examples/loader-rust-to-json/src/data/poker.json.rs:1]: Yeah this is really cool

File: examples/loader-rust-to-json/src/index.md (hunk 7 above)
Comment (mythmon) [reply examples/loader-rust-to-json/src/index.md:201]:
```suggestion
We can make a quick chart of the poker hands with Plot:
```

**Review by allisonhorst — APPROVED** [review]
